Fixed the issue in serialization of emoji chars

crimsonred opened the pull request:
On serializing a text with emoji in it (Text with 😜 emoji 🎉), using the code (running on Mono framework ver 3.2.5):

```
var writer = new JsonFx.Json.JsonWriter ();
string json = writer.Write (objectToSerialize);
```

the library threw an exception: "The string contains invalid surrogate pair character at….". 

This has been fixed in this pull request.

mckamey commented:
This looks like a good change but can you clean up the whitespace to be consistent with the original?

crimsonred commented:
@mckamey, my apologies for the non-consistent whitespacing, my editor was set to replace tabs with 4 spaces.

mckamey commented:
Sorry I should have noticed this when I commented about the whitespace. The change you're making seems to make more sense inside the CharUtility.ConvertToUtf32 method. It already has the Silverlight handling and just needs the surrogate check added:

  if (Char.IsSurrogate(value, index)) {
      return (int)value[index];
  }

https://github.com/jsonfx/jsonfx/blob/master/src/JsonFx/Utils/CharUtility.cs#L172

If you wouldn't mind move it over there and reverting `JsonFormatter.cs`. Let me know if it still solves your emoji situation.

crimsonred commented:
@mckamey, I remember I did try to add it in CharUtility.cs but I don't remember the issues I faced. Let me try it again, I will get back to you in a day or two.

mckamey commented:
@crimsonred, if you add a new unit test to [JsonFormatterTests.cs](https://github.com/jsonfx/jsonfx/blob/master/src/JsonFx.Tests/Json/JsonFormatterTests.cs) then you can easily demonstrate the situation that fails and that the fix fixes it. Add the unit test and it should be failing. Add the fix and it should pass.

crimsonred commented:
@mckamey, I have moved the changes and added the unit test. If you remove the changes the unit test fails with an exception: "The string contains invalid surrogate pair character at….".

mckamey commented on src/JsonFx/Json/JsonFormatter.cs:
```diff
@@ -512,11 +512,11 @@ protected virtual void WriteNaN(TextWriter writer)
 			{
 				int start = 0,
-					length = value.Length;
+ 				length = value.Length;
```
Can you revert this file since this is only changing whitespace.

mckamey commented on src/JsonFx/Utils/CharUtility.cs:
```diff
@@ -169,7 +169,14 @@ public static int ConvertToUtf32(string value, int index)
 			return (int)value[index];
 #else
-			return Char.ConvertToUtf32(value, index);
+			if (char.IsSurrogate(value[index]))
```
how about:

    if (Char.IsSurrogate(value, index))

mckamey commented:
Nice. Thanks for adding the unit test. I made two comments and then I'll merge.

crimsonred commented:
@mckamey, I have modified the files as per your suggestions.

mckamey commented:
Thanks for making those updates.
